sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2OptionSuite.scala

```diff
@@ -0,0 +1,290 @@
+          val append = write.toBatch.asInstanceOf[InMemoryBaseTable#Append]
+          assert(append.info.options.get("write.split-size") === "10")
+      }
+      assert (collected.size == 1)
+
+      val insertResult = sql(s"SELECT * FROM $t1")
+      checkAnswer(insertResult, Seq(Row(1, "a"), Row(2, "b")))
+    }
+  }
+
+  test("SPARK-50286: Propagate options for DataFrame Append") {
+    val t1 = s"${catalogAndNamespace}table"
+    withTable(t1) {
+      sql(s"CREATE TABLE $t1 (id bigint, data string)")
+      withListener {
+        sql("VALUES (1, 'a'), (2, 'b')")
+          .withColumnRenamed("col1", "id")
+          .withColumnRenamed("col2", "data")
+          .writeTo(t1)
+          .option("write.split-size", "10")
+          .append()
+      } { qe =>
+        var collected = qe.optimizedPlan.collect {
+          case AppendData(_: DataSourceV2Relation, _, writeOptions, _, _, _) =>
+            assert(writeOptions("write.split-size") == "10")
+        }
+        assert (collected.size == 1)
+
+        collected = qe.executedPlan.collect {
+          case AppendDataExec(_, _, write) =>
+            val append = write.toBatch.asInstanceOf[InMemoryBaseTable#Append]
+            assert(append.info.options.get("write.split-size") === "10")
+        }
+        assert (collected.size == 1)
+      }
+    }
+  }
+
+  test("SPARK-49098, SPARK-50286: Supports Dynamic Table Options for SQL Insert Overwrite") {
+    val t1 = s"${catalogAndNamespace}table"
+    withTable(t1) {
+      sql(s"CREATE TABLE $t1 (id bigint, data string)")
+      sql(s"INSERT INTO $t1 VALUES (1, 'a'), (2, 'b')")
+
+      val df = sql(s"INSERT OVERWRITE $t1 WITH (`write.split-size` = 10) " +
+        s"VALUES (3, 'c'), (4, 'd')")
+      var collected = df.queryExecution.optimizedPlan.collect {
+        case CommandResult(_,
+          OverwriteByExpression(relation: DataSourceV2Relation, _, _, _, _, _, _),
+          _, _) =>
+          assert(relation.options.get("write.split-size") === "10")
+      }
+      assert (collected.size == 1)
+
+      collected = df.queryExecution.executedPlan.collect {
+        case CommandResultExec(
+          _, OverwriteByExpressionExec(_, _, write),
+          _) =>
+          val append = write.toBatch.asInstanceOf[InMemoryBaseTable#TruncateAndAppend]
+          assert(append.info.options.get("write.split-size") === "10")
+      }
+      assert (collected.size == 1)
+
+      val insertResult = sql(s"SELECT * FROM $t1")
+      checkAnswer(insertResult, Seq(Row(3, "c"), Row(4, "d")))
+    }
+  }
+
+  test("SPARK-50286: Propagate options for DataFrame OverwritePartitions") {
+    val t1 = s"${catalogAndNamespace}table"
+    withTable(t1) {
+      sql(s"CREATE TABLE $t1 (id bigint, data string)")
+      sql(s"INSERT INTO $t1 VALUES (1, 'a'), (2, 'b')")
+
+      withListener {
+        sql("VALUES (3, 'c'), (4, 'd')")
+          .withColumnRenamed("col1", "id")
+          .withColumnRenamed("col2", "data")
+          .writeTo(t1)
```

Comment:
shall we also test the v1 `DataFrameWriter` API?

Reply:
sure, add two cases
- `SPARK-50286: Propagate options for DataFrameWriter Append`
- `SPARK-50286: Propagate options for DataFrameWriter Overwrite`